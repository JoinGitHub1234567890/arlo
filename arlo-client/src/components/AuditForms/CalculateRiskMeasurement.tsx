import React from 'react'
import styled from 'styled-components'
import { toast } from 'react-toastify'
import { Formik, FormikProps, FieldArray, Form, Field } from 'formik'
// import * as Yup from 'yup'
import FormSection, {
  FormSectionLabel,
  FormSectionDescription,
} from '../Form/FormSection'
import FormWrapper from '../Form/FormWrapper'
import FormButton from '../Form/FormButton'
import FormField from '../Form/FormField'
import FormButtonBar from '../Form/FormButtonBar'
import { api } from '../utilities'
import { Contest, Round, Candidate, RoundContest } from '../../types'

const InputSection = styled.div`
  display: block;
  margin-top: 25px;
  width: 100%;
  font-size: 0.4em;
`

const InputLabel = styled.label`
  display: inline-block;
`

const InlineInput = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin-bottom: 10px;
  width: 50%;
`

interface Props {
  audit: any
  isLoading: boolean
  setIsLoading: (isLoading: boolean) => void
  updateAudit: () => void
}

interface CalculateRiskMeasurementValues {
  round: number
  contests: {
    [key: string]: number | ''
  }[]
}

type AggregateContest = Contest & RoundContest

const CalculateRiskMeasurmeent = (props: Props) => {
  const { audit, isLoading, setIsLoading, updateAudit } = props

  //const sumOfAuditedVotes: { current: number } = useRef(0)

  const downloadBallotRetrievalList = (id: number, e: any) => {
    e.preventDefault()
    const jurisdictionID: string = audit.jurisdictions[0].id
    window.open(`/jurisdiction/${jurisdictionID}/${id}/retrieval-list`)
  }

  const downloadAuditReport = async (e: React.MouseEvent) => {
    e.preventDefault()
    try {
      window.open(`/audit/report`)
      updateAudit()
    } catch (err) {
      toast.error(err.message)
    }
  }

  const calculateRiskMeasurement = async (
    values: CalculateRiskMeasurementValues
  ) => {
    try {
      const jurisdictionID: string = audit.jurisdictions[0].id
      const body: any = {
        contests: audit.contests.map((contest: Contest, i: number) => ({
          // map values.contests
          id: contest.id,
          results: {
            ...values.contests[i],
          },
        })),
      }

      setIsLoading(true)
      await api(`/jurisdiction/${jurisdictionID}/${values.round}/results`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      })
      //sumOfAuditedVotes.current +=
      //  Number(values['candidate-1']) + Number(values['candidate-2'])
      updateAudit()
    } catch (err) {
      toast.error(err.message)
    }
  }

  const rounds: CalculateRiskMeasurementValues[] = audit.rounds.map(
    (r: Round, i: number) => ({
      contests: audit.contests.map((contest: Contest, j: number) => ({
        ...contest.choices.reduce((acc, choice: Candidate) => {
          return { ...acc, [choice.id]: r.contests[j].results[choice.id] || '' }
        }, {}),
      })),
      round: i + 1,
    })
  )

  return audit.rounds.map((round: Round, i: number) => {
    const aggregateContests: AggregateContest[] = audit.contests.reduce(
      (acc: any[], contest: Contest) => {
        const roundContest = round.contests.find(v => v.id === contest.id)
        acc.push({ ...contest, ...roundContest })
        return acc
      },
      []
    )
    const completeContests = aggregateContests.reduce((acc, c) => {
      if (c.endMeasurements.isComplete) {
        acc += 1
      }
      return acc
    }, 0)
    /*
      const overCount = function(this: {
      parent: { [key: string]: number }
    }): boolean {
      const count = Object.values(this.parent).reduce(
        (c: number, v: number) => {
          c += v || 0
          return c
        },
        0
      )
      const maxVotes = this.parent.parent.id
      return count + sumOfAuditedVotes.current <= max
    }
    const schema = Yup.object().shape({
      contests: Yup.lazy(() => {
        return Object.keys(v.contests).reduce((acc: any, result: string) => {
          acc[result] = Yup.number().test(
            'overCount',
            'Cannot exceed the number of total ballots cast',
            overCount
          )
          return acc
        }, {})
      })
    }) */
    /* eslint-disable react/no-array-index-key */
    const aggregatedBallots = aggregateContests.reduce(
      (acc: number, contest: AggregateContest) => {
        acc += contest.sampleSize
        return acc
      },
      0
    )
    return (
      <Formik
        key={i}
        onSubmit={calculateRiskMeasurement}
        initialValues={rounds[i]}
        //validationSchema={schema}
        enableReinitialize
        render={({
          values,
          handleSubmit,
        }: FormikProps<CalculateRiskMeasurementValues>) => (
          <Form>
            <FormWrapper title={`Round ${i + 1}`}>
              <FormSectionLabel>
                Ballot Retrieval List: {aggregatedBallots} Total Ballots
              </FormSectionLabel>
              <FormSectionDescription>
                {aggregateContests.map(
                  (contest: AggregateContest, i: number) => (
                    <p key={contest.id}>
                      Contest {i + 1}: {contest.sampleSize} ballots
                    </p>
                  )
                )}
              </FormSectionDescription>
              {/*<SectionLabel>
                Ballot Retrieval List \n
                {contest ? `${contest.sampleSize} Ballots` : ''}
              </SectionLabel>*/}
              <FormButton
                onClick={(e: React.MouseEvent) =>
                  downloadBallotRetrievalList(i + 1, e)
                }
                inline
              >
                Download Aggregated Ballot Retrieval List for Round {i + 1}
              </FormButton>
              {/** contest iteration */}
              <FieldArray
                name="contests"
                render={() => {
                  return (
                    <>
                      {values.contests.map((contest, j) => {
                        return (
                          <FormSection
                            key={aggregateContests[j].id}
                            label={`Contest ${j + 1}: ${
                              aggregateContests[j].name
                            }`}
                          >
                            {aggregateContests[j].endMeasurements.isComplete ? (
                              <>
                                <FormSectionLabel>
                                  Contest Status: COMPLETE
                                </FormSectionLabel>
                                <InputSection>
                                  <InlineInput>
                                    <InputLabel>Risk Limit: </InputLabel>
                                    {audit.riskLimit}%
                                  </InlineInput>
                                  <InlineInput>
                                    <InputLabel>P-value: </InputLabel>{' '}
                                    {
                                      aggregateContests[j].endMeasurements
                                        .pvalue
                                    }
                                  </InlineInput>
                                </InputSection>
                              </>
                            ) : (
                              <>
                                <FormSectionLabel>
                                  Audited Results: Round {i + 1}, Contest{' '}
                                  {j + 1} INCOMPLETE
                                </FormSectionLabel>
                                <FormSectionDescription>
                                  Enter the number of votes recorded for each
                                  candidate/choice in the audited ballots for
                                  Round {i + 1}, Contest {j + 1}
                                </FormSectionDescription>
                                <InputSection>
                                  {Object.keys(contest).map(choiceId => {
                                    const name = aggregateContests[
                                      j
                                    ].choices.find(
                                      (candidate: Candidate) =>
                                        candidate.id === choiceId
                                    )!.name
                                    return (
                                      <React.Fragment key={choiceId}>
                                        <InlineInput>
                                          <InputLabel>{name}</InputLabel>
                                          <Field
                                            name={`contests[${j}][${choiceId}]`}
                                            type="number"
                                            component={FormField}
                                          />
                                        </InlineInput>
                                      </React.Fragment>
                                    )
                                  })}
                                </InputSection>
                              </>
                            )}
                          </FormSection>
                        )
                      })}
                    </>
                  )
                }}
              />
              {isLoading && <p>Loading...</p>}
              <FormSection>
                <FormSectionLabel>
                  Audit Progress: {completeContests} of {audit.contests.length}{' '}
                  complete
                </FormSectionLabel>
              </FormSection>
              {i + 1 === audit.rounds.length &&
                aggregateContests.some(
                  (contest: AggregateContest) =>
                    !contest.endMeasurements.isComplete
                ) &&
                !isLoading && (
                  <FormButtonBar>
                    <FormButton type="button" onClick={handleSubmit}>
                      Calculate Risk Measurement
                    </FormButton>
                  </FormButtonBar>
                )}
              <FormSection>
                {/*<FormSectionLabel>
                  Audit Status:{' '}
                  {contest.endMeasurements.isComplete
                    ? 'COMPLETE'
                    : 'INCOMPLETE'}
                </FormSectionLabel>
                <InputSection>
                  <InlineInput>
                    <InputLabel>Risk Limit: </InputLabel>
                    {audit.riskLimit}%
                  </InlineInput>
                  <InlineInput>
                    <InputLabel>P-value: </InputLabel>{' '}
                    {contest.endMeasurements.pvalue}
                  </InlineInput>
                </InputSection>*/}
                {/* {Form 3} */}
                {completeContests === audit.contests.length && (
                  <FormButton
                    onClick={(e: React.MouseEvent) => downloadAuditReport(e)}
                    size="sm"
                    inline
                  >
                    Download Audit Report
                  </FormButton>
                )}
              </FormSection>
            </FormWrapper>
          </Form>
        )}
      />
    )
  })
}

export default React.memo(CalculateRiskMeasurmeent)
