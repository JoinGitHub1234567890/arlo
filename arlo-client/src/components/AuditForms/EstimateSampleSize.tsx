import React from 'react'
import styled from 'styled-components'
import { toast } from 'react-toastify'
import {
  Formik,
  FormikProps,
  Form,
  Field,
  FieldArray,
  ErrorMessage,
} from 'formik'
import * as Yup from 'yup'
import uuidv4 from 'uuidv4'
import FormSection from '../Form/FormSection'
import FormWrapper from '../Form/FormWrapper'
import FormTitle from '../Form/FormTitle'
import FormButton from '../Form/FormButton'
import FormField from '../Form/FormField'
import FormButtonBar from '../Form/FormButtonBar'
// import { Audit } from '../../types'
import { api } from '../utilities'
import { generateOptions, ErrorLabel } from '../Form/_helpers'

const TwoColumnSection = styled.div`
  display: block;
  margin-top: 25px;
  width: 100%;
  font-size: 0.4em;
`

const InputLabelRow = styled.div`
  display: flex;
  flex-direction: row;
  margin-bottom: 10px;
  width: 100%;
`
const InputFieldRow = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  margin-bottom: 25px;
  width: 100%;
`

const FieldLeft = styled(FormField)`
  flex-grow: 2;
  width: unset;
`

const FieldRight = styled(FieldLeft)`
  margin-left: 50px;
`

const InputLabel = styled.label`
  display: inline-block;
`

const InputLabelRight = styled.label`
  margin-left: 75px;
`

const Action = styled.p`
  margin: 5px 0 0 0;
  width: 100%;
  color: #000088;
  font-size: 14px;
  &:hover {
    cursor: pointer;
  }
`

interface Props {
  audit?: any
  isLoading?: any
  setIsLoading: (isLoading: boolean) => void
  updateAudit: () => void
}

interface ChoiceValues {
  id: number
  name: string
  numVotes: number
}

interface ContestValues {
  name?: string
  totalBallotsCast?: number
  choices: ChoiceValues[]
}

interface EstimateSampleSizeValues {
  name: string
  randomSeed: number
  riskLimit: number
  contests: ContestValues[]
}

const contestsSchema = Yup.array()
  .required()
  .of(
    Yup.object().shape({
      name: Yup.string()
        .min(2, 'Name must be longer than 2 characters')
        .max(50, 'Name must be shorter than 50 characters')
        .required('Required'),
      totalBallotsCast: Yup.number().required('Required'),
      choices: Yup.array()
        .required()
        .of(
          Yup.object().shape({
            name: Yup.string()
              .required('Required')
              .min(2, 'Name must be longer than 2 characters')
              .max(50, 'Name must be shorter than 50 characters'),
            numVotes: Yup.number().required('Required'),
          })
        ),
    })
  )

const schema = Yup.object().shape({
  name: Yup.string().required('Required'),
  randomSeed: Yup.number()
    .min(1, 'Must be at least 1')
    .max(99999999999999999999, 'Cannot exceed 20 digits')
    .required('Required'),
  riskLimit: Yup.number()
    .min(1, 'Must be greater than 0')
    .max(20, 'Must be less than 21')
    .required('Required'),
  contests: contestsSchema,
})

const EstimateSampleSize = ({
  audit,
  isLoading,
  setIsLoading,
  updateAudit,
}: Props) => {
  const canEstimateSampleSize = !audit.contests.length

  const handlePost = async (values: EstimateSampleSizeValues) => {
    const data = {
      // incomplete Audit
      name: values.name,
      randomSeed: Number(values.randomSeed),
      riskLimit: Number(values.riskLimit),
      contests: values.contests.map(contest => ({
        id: uuidv4(),
        name: contest.name,
        totalBallotsCast: Number(contest.totalBallotsCast),
        choices: contest.choices.map(choice => ({
          id: uuidv4(),
          name: choice.name,
          numVotes: Number(choice.numVotes),
        })),
      })),
    }
    try {
      setIsLoading(true)
      await api(`/audit/basic`, {
        method: 'POST',
        body: JSON.stringify(data),
        headers: {
          'Content-Type': 'application/json',
        },
      })
      updateAudit()
    } catch (err) {
      toast.error(err.message)
    } finally {
      setIsLoading(false)
    }
  }

  const contestValues = [
    {
      name: '',
      totalBallotsCast: '',
      choices: [
        {
          name: '',
          numVotes: '',
        },
        {
          name: '',
          numVotes: '',
        },
      ],
    },
  ]

  const initialValues = {
    randomSeed: audit.randomSeed || '',
    riskLimit: audit.riskLimit || 1,
    name: audit.name || '',
    contests: audit.contests.length ? audit.contests : contestValues,
  }

  return (
    <>
      <h1>Audit Setup</h1>
      <Formik
        initialValues={initialValues}
        validationSchema={schema}
        onSubmit={handlePost}
        enableReinitialize
      >
        {({ values, handleSubmit }: FormikProps<EstimateSampleSizeValues>) => (
          <Form>
            <FormWrapper>
              <FormSection label="Audit Name">
                <Field name="name" component={FormField} />
              </FormSection>
              <FieldArray
                name="contests"
                render={contestsArrayHelpers => (
                  <>
                    {values.contests.map(
                      (contest: ContestValues, i: number) => (
                        /* eslint-disable react/no-array-index-key */
                        <React.Fragment key={i}>
                          {i > 0 && (
                            <FormSection>
                              <hr />
                            </FormSection>
                          )}
                          <FormSection
                            label={`Contest ${
                              values.contests.length > 1 ? i + 1 : ''
                            } Name`}
                            description="Enter the name of the contest that will drive the audit."
                          >
                            <Field
                              name={`contests[${i}].name`}
                              disabled={!canEstimateSampleSize}
                              component={FormField}
                            />
                            {values.contests.length > 1 &&
                              !audit.jurisdictions.length && (
                                <Action
                                  onClick={() => contestsArrayHelpers.remove(i)}
                                >
                                  Remove Contest {i + 1}
                                </Action>
                              )}
                          </FormSection>
                          <FieldArray
                            name={`contests[${i}].choices`}
                            render={choicesArrayHelpers => (
                              <FormSection
                                label="Candidates/Choices & Vote Totals"
                                description="Enter the name of each candidate choice that appears on the ballot for this contest."
                              >
                                <TwoColumnSection>
                                  {contest.choices.map(
                                    (choice: ChoiceValues, j: number) => (
                                      /* eslint-disable react/no-array-index-key */
                                      <React.Fragment key={j}>
                                        <InputLabelRow>
                                          <InputLabel>
                                            Name of Candidate/Choice {j + 1}
                                          </InputLabel>
                                          <InputLabelRight>
                                            Votes for Candidate/Choice {j + 1}
                                          </InputLabelRight>
                                        </InputLabelRow>
                                        <InputFieldRow>
                                          <Field
                                            name={`contests[${i}].choices[${j}].name`}
                                            disabled={!canEstimateSampleSize}
                                            component={FieldLeft}
                                          />
                                          <Field
                                            name={`contests[${i}].choices[${j}].numVotes`}
                                            type="number"
                                            disabled={!canEstimateSampleSize}
                                            component={FieldRight}
                                          />
                                          {contest.choices.length > 2 &&
                                            !audit.jurisdictions.length && (
                                              <Action
                                                onClick={() =>
                                                  choicesArrayHelpers.remove(j)
                                                }
                                              >
                                                Remove choice {j + 1}
                                              </Action>
                                            )}
                                        </InputFieldRow>
                                      </React.Fragment>
                                    )
                                  )}
                                  {!audit.jurisdictions.length && (
                                    <Action
                                      onClick={() =>
                                        choicesArrayHelpers.push({
                                          name: '',
                                          numVotes: '',
                                        })
                                      }
                                    >
                                      Add a new candidate/choice
                                    </Action>
                                  )}
                                </TwoColumnSection>
                              </FormSection>
                            )}
                          />
                          <FormSection
                            label="Total Ballots Cast"
                            description="Enter the overall number of ballot cards cast in jurisdictions containing this contest."
                          >
                            <Field
                              type="number"
                              name={`contests[${i}].totalBallotsCast`}
                              disabled={!canEstimateSampleSize}
                              component={FormField}
                            />
                          </FormSection>
                        </React.Fragment>
                      )
                    )}
                    <FormButtonBar>
                      {!audit.jurisdictions.length && (
                        <FormButton
                          type="button"
                          onClick={() =>
                            contestsArrayHelpers.push({ ...contestValues[0] })
                          }
                        >
                          Add another targeted contest
                        </FormButton>
                      )}
                    </FormButtonBar>
                  </>
                )}
              />
              <FormTitle>Audit Settings</FormTitle>
              <FormSection
                label="Desired Risk Limit"
                description='Set the risk for the audit as as percentage (e.g. "5" = 5%'
              >
                <Field
                  name="riskLimit"
                  disabled={!canEstimateSampleSize}
                  component="select"
                >
                  {generateOptions(20)}
                </Field>
                <ErrorMessage name="riskLimit" component={ErrorLabel} />
              </FormSection>
              <FormSection
                label="Random Seed"
                description="Enter the random number to seed the pseudo-random number generator."
              >
                <Field
                  type="number"
                  name="randomSeed"
                  disabled={!canEstimateSampleSize}
                  component={FormField}
                />
              </FormSection>
            </FormWrapper>
            {!audit.contests.length && isLoading && <p>Loading...</p>}
            {!audit.contests.length && !isLoading && (
              <FormButtonBar>
                <FormButton
                  type="submit"
                  disabled={!canEstimateSampleSize}
                  onClick={handleSubmit}
                >
                  Estimate Sample Size
                </FormButton>
              </FormButtonBar>
            )}
          </Form>
        )}
      </Formik>

      {/** old code
      {Array.from(Array(numContests).keys()).map(i => {
        return (
          <Formik
            key={i}
            initialValues={contestValues[i]}
            validationSchema={contestsSchema}
            onSubmit={values => {
              contests.current.push(values)
            }}
            enableReinitialize
            render={({
              values,
              errors,
              touched,
              handleChange,
              handleBlur,
              submitForm,
              handleSubmit,
            }: FormikProps<ContestValues>) => {
              contestForms.current[i] = submitForm
              return (
                <form onSubmit={handleSubmit}>
                  <FormWrapper title="Contest Information">
                    <React.Fragment key={i}>
                      {i > 0 && (
                        <FormSection>
                          <hr />
                        </FormSection>
                      )}
                      <FormSection
                        label={`Contest ${numContests > 1 ? i + 1 : ''} Name`}
                        description="Enter the name of the contest that will drive the audit."
                      >
                        <FormField
                          name="name"
                          value={values.name}
                          onChange={handleChange}
                          onBlur={handleBlur}
                          disabled={!canEstimateSampleSize}
                          error={errors.name}
                          touched={touched.name}
                        />
                      </FormSection>
                      <FormSection
                        label="Candidates/Choices & Vote Totals"
                        description="Enter the name of each candidate choice that appears on the ballot for this contest."
                      >
                        <TwoColumnSection>
                          <InputLabelRow>
                            <InputLabel>Name of Candidate/Choice 1</InputLabel>
                            <InputLabelRight>
                              Votes for Candidate/Choice 1
                            </InputLabelRight>
                          </InputLabelRow>
                          <InputFieldRow>
                            <FormField
                              name="candidateOneName"
                              value={values.candidateOneName}
                              onChange={handleChange}
                              onBlur={handleBlur}
                              disabled={!canEstimateSampleSize}
                              error={errors.candidateOneName}
                              touched={touched.candidateOneName}
                            />
                            <FieldRight
                              type="number"
                              name="candidateOneVotes"
                              onChange={handleChange}
                              onBlur={handleBlur}
                              value={values.candidateOneVotes}
                              disabled={!canEstimateSampleSize}
                              error={errors.candidateOneVotes}
                              touched={touched.candidateOneVotes}
                            />
                          </InputFieldRow>
                          <InputLabelRow>
                            <InputLabel>Name of Candidate/Choice 2</InputLabel>
                            <InputLabelRight>
                              Votes for Candidate/Choice 2
                            </InputLabelRight>
                          </InputLabelRow>
                          <InputFieldRow>
                            <FormField
                              name="candidateTwoName"
                              onChange={handleChange}
                              onBlur={handleBlur}
                              value={values.candidateTwoName}
                              disabled={!canEstimateSampleSize}
                              error={errors.candidateTwoName}
                              touched={touched.candidateTwoName}
                            />
                            <FieldRight
                              type="number"
                              name="candidateTwoVotes"
                              onChange={handleChange}
                              onBlur={handleBlur}
                              value={values.candidateTwoVotes}
                              disabled={!canEstimateSampleSize}
                              error={errors.candidateTwoVotes}
                              touched={touched.candidateTwoVotes}
                            />
                          </InputFieldRow>
                        </TwoColumnSection>
                      </FormSection>

                      <FormSection
                        label="Total Ballots Cast"
                        description="Enter the overall number of ballot cards cast in jurisdictions containing this contest."
                      >
                        <FormField
                          type="number"
                          name="totalBallotsCast"
                          onChange={handleChange}
                          onBlur={handleBlur}
                          value={values.totalBallotsCast}
                          disabled={!canEstimateSampleSize}
                          error={errors.totalBallotsCast}
                          touched={
                            touched.totalBallotsCast &&
                            touched.candidateOneVotes &&
                            touched.candidateTwoVotes
                          }
                        />
                      </FormSection>
                    </React.Fragment>
                  </FormWrapper>
                </form>
              )
            }}
          />
        )
      })}
      <Formik
        initialValues={initialValues}
        validationSchema={schema}
        onSubmit={handlePost}
        enableReinitialize
        render={({
          values,
          errors,
          touched,
          handleChange,
          handleBlur,
          handleSubmit,
        }: FormikProps<EstimateSampleSizeValues>) => (
          <form>
            <FormWrapper>
              <FormTitle>Audit Settings</FormTitle>
              <FormSection
                label="Desired Risk Limit"
                description='Set the risk for the audit as as percentage (e.g. "5" = 5%'
              >
                <select
                  name="riskLimit"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.riskLimit}
                  disabled={!canEstimateSampleSize}
                >
                  {generateOptions(20)}
                </select>
                {errors.riskLimit && touched.riskLimit && (
                  <ErrorLabel>{errors.riskLimit}</ErrorLabel>
                )}
              </FormSection>
              <FormSection
                label="Random Seed"
                description="Enter the random number to seed the pseudo-random number generator."
              >
                <FormField
                  type="number"
                  onChange={handleChange}
                  value={values.randomSeed}
                  onBlur={handleBlur}
                  name="randomSeed"
                  disabled={!canEstimateSampleSize}
                  error={errors.randomSeed}
                  touched={touched.randomSeed}
                />
              </FormSection>
            </FormWrapper>
            <FormButtonBar>
              <FormButton type="button" onClick={addContest}>
                Add another targeted contest
              </FormButton>
            </FormButtonBar>
            {!audit.contests.length && isLoading && <p>Loading...</p>}
            {!audit.contests.length && !isLoading && (
              <FormButtonBar>
                <FormButton
                  type="submit"
                  disabled={!canEstimateSampleSize}
                  onClick={handleSubmit}
                >
                  Estimate Sample Size
                </FormButton>
              </FormButtonBar>
            )}
          </form>
        )}
      />
      */}
    </>
  )
}

export default React.memo(EstimateSampleSize)
