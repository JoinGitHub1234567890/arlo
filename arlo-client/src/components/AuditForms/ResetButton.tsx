import React from 'react'
import ReactDOM from 'react-dom'
import styled from 'styled-components'
import { api } from '../utilities'

const Button = styled.button`
  margin: 0 auto;
  border-radius: 5px;
  background: rgb(211, 211, 211);
  width: 155px;
  height: 30px;
  color: #000000;
  font-size: 0.4em;
  font-weight: 500;
`

interface Props {
  electionId: string
  disabled?: boolean
  history: any
}

const ResetButton: React.FC<Props> = ({
  electionId,
  disabled,
  history,
}: Props) => {
  const resetButtonWrapper = document.getElementById('reset-button-wrapper')
  const reset = async () => {
    await api(`/audit/reset`, { electionId, method: 'POST' })

    history.push('/')
  }
  if (resetButtonWrapper) {
    return ReactDOM.createPortal(
      <Button onClick={reset} disabled={disabled}>
        Clear &amp; Restart
      </Button>,
      resetButtonWrapper
    )
  }
  return null
}

export default ResetButton
