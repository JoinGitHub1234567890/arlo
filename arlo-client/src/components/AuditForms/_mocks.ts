import { Audit } from '../../types'

export const mockAudit: Audit = {
  name: 'Primary 2019',
  riskLimit: 10,
  randomSeed: 123123123,

  contests: [
    {
      id: 'contest-1',
      name: 'Contest 1',

      choices: [
        {
          id: 'candidate-1',
          name: 'Candidate 1',
          numVotes: 42,
        },
        {
          id: 'candidate-2',
          name: 'Candidate 2',
          numVotes: 72,
        },
      ],

      totalBallotsCast: 4200,
    },
  ],

  jurisdictions: [
    {
      id: 'adams-county',
      name: 'Adams County',
      contests: ['contest-1'],
      auditBoards: [
        {
          id: 'audit-board-1',
          members: [],
        },
        {
          id: 'audit-board-2',
          members: [],
        },
      ],
      ballotManifest: {
        filename: 'Adams_County_Manifest.csv',
        numBallots: 123456,
        numBatches: 560,
        uploadedAt: '2019-06-17 11:45:00',
      },
    },
  ],

  rounds: [
    {
      startedAt: '2019-06-17 11:45:00',
      endedAt: '2019-06-17 11:55:00',
      contests: [
        {
          id: 'contest-1',
          endMeasurements: {
            pvalue: 0.085,
            isComplete: false,
          },
          results: {
            'candidate-1': 55,
            'candidate-2': 35,
          },
          sampleSize: 25,
        },
      ],
      jurisdictions: {
        'adams-county': {
          numBallots: 15,
        },
      },
    },
  ],
}

export default mockAudit
