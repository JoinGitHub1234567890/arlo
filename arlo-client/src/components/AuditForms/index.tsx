import React, { useState, useEffect, useCallback } from 'react'
import EstimateSampleSize from './EstimateSampleSize'
import SelectBallotsToAudit from './SelectBallotsToAudit'
import CalculateRiskMeasurement from './CalculateRiskMeasurement'
import { api } from '../utilities'
import { Audit, Params } from '../../types'
import ResetButton from './ResetButton'

const initialData: Audit = {
  name: '',
  riskLimit: '',
  randomSeed: '',
  contests: [],
  jurisdictions: [],
  rounds: [],
}

const AuditForms: React.FC<any> = ({
  match: {
    params: { electionId },
  },
  history,
}: {
  match: { params: Params }
  history: any
}) => {
  const [isLoading, setIsLoading] = useState<boolean>(false)

  const [audit, setAudit] = useState(initialData)

  const getStatus = useCallback(async (): Promise<Audit> => {
    const audit: Audit = await api('/audit/status', { electionId })
    return audit
  }, [electionId])

  const updateAudit = useCallback(async () => {
    const audit = await getStatus()
    setIsLoading(true)
    setAudit(audit)
    setIsLoading(false)
  }, [getStatus])

  useEffect(() => {
    updateAudit()
  }, [updateAudit])

  return (
    <React.Fragment>
      <ResetButton
        electionId={electionId}
        disabled={!audit.contests.length}
        history={history}
      />

      <EstimateSampleSize
        audit={audit}
        isLoading={isLoading}
        setIsLoading={setIsLoading}
        updateAudit={updateAudit}
        electionId={electionId}
      />

      {!!audit.contests.length && (
        <SelectBallotsToAudit
          audit={audit}
          isLoading={isLoading}
          setIsLoading={setIsLoading}
          updateAudit={updateAudit}
          getStatus={getStatus}
          electionId={electionId}
        />
      )}

      {!!audit.rounds.length && (
        <CalculateRiskMeasurement
          audit={audit}
          isLoading={isLoading}
          setIsLoading={setIsLoading}
          updateAudit={updateAudit}
          electionId={electionId}
        />
      )}
    </React.Fragment>
  )
}

export default AuditForms
