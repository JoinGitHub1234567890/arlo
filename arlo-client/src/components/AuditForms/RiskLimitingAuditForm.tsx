import React, { useState, useEffect, useCallback } from 'react'
import EstimateSampleSize from './EstimateSampleSize'
import SelectBallotsToAudit from './SelectBallotsToAudit'
import CalculateRiskMeasurement from './CalculateRiskMeasurement'
//import { api } from '../utilities'
import { Audit } from '../../types'
import { statusStates } from './_mocks'

const AuditForms = () => {
  const [isLoading, setIsLoading] = useState<boolean>(false)

  const [audit, setAudit] = useState<Audit>(statusStates[0])

  const getStatus = useCallback(async (): Promise<Audit> => {
    //const audit: any = await api('/audit/status', {})
    return statusStates[4]
  }, [])

  const updateAudit = useCallback(async () => {
    const audit = await getStatus()
    setIsLoading(true)
    setAudit(audit)
    setIsLoading(false)
  }, [getStatus])

  useEffect(() => {
    updateAudit()
  }, [updateAudit])

  return (
    <React.Fragment>
      <EstimateSampleSize
        audit={audit}
        isLoading={isLoading}
        setIsLoading={setIsLoading}
        updateAudit={updateAudit}
      />

      {!!audit.contests.length && (
        <SelectBallotsToAudit
          audit={audit}
          isLoading={isLoading}
          setIsLoading={setIsLoading}
          updateAudit={updateAudit}
          getStatus={getStatus}
        />
      )}

      {!!audit.rounds.length && (
        <CalculateRiskMeasurement
          audit={audit}
          isLoading={isLoading}
          setIsLoading={setIsLoading}
          updateAudit={updateAudit}
        />
      )}
    </React.Fragment>
  )
}

export default AuditForms
