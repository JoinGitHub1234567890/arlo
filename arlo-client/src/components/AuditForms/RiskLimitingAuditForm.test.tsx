import React from 'react'
// import { render } from '@testing-library/react'
import { shallow } from 'enzyme'
import AuditForms from './RiskLimitingAuditForm'

const mockAudit = {
  name: 'Primary 2019',
  riskLimit: 10,
  randomSeed: 'sdfkjsdflskjfd',

  contests: [
    {
      id: 'contest-1',
      name: 'Contest 1',

      choices: [
        {
          name: 'Candidate 1',
          numVotes: 42,
        },
      ],

      totalBallotsCast: 4200,
    },
  ],

  jurisdictions: [
    {
      id: 'adams-county',
      name: 'Adams County',
      contests: ['contest-1'],
      auditBoards: [
        {
          id: 'audit-board-1',
          members: [],
        },
        {
          id: 'audit-board-2',
          members: [],
        },
      ],
      ballotManifest: {
        filename: 'Adams_County_Manifest.csv',
        numBallots: 123456,
        numBatches: 560,
        uploadedAt: '2019-06-17 11:45:00',
      },
    },
  ],

  rounds: [
    {
      startedAt: '2019-06-17 11:45:00',
      endedAt: '2019-06-17 11:55:00',
      contests: [
        {
          id: 'contest-1',
          endMeasurements: {
            pvalue: 0.085,
            isComplete: false,
          },
          results: {
            'candidate-1': 55,
            'candidate-2': 35,
          },
          sampleSize: 25,
        },
      ],
      jurisdictions: {
        'adams-county': {
          numBallots: 15,
        },
      },
    },
  ],
}

it('renders corretly', () => {
  const container = shallow(<AuditForms />)
  expect(container).toMatchSnapshot()
  container.setState({ mockAudit })
  expect(container).toMatchSnapshot()
})
