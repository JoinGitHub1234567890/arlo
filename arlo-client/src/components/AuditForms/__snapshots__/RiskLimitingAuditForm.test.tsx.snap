// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders corretly 1`] = `ShallowWrapper {}`;

exports[`renders corretly 2`] = `ShallowWrapper {}`;
